src/libraries/Common/src/Interop/OSX/Swift.Runtime/Foundation.cs

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#pragma warning disable CS3016 // Arrays as attribute arguments are not CLS Compliant
+#pragma warning disable SYSLIB1051
+
+namespace Swift.Runtime
+{
+    /// <summary>
+    /// Represents a Swift type in C#.
+    /// </summary>
+    internal unsafe interface ISwiftType
+    {
+        public abstract void* Metadata();
+    }
+
+    /// <summary>
+    /// Represents a Swift protocol in C#.
+    /// </summary>
+    internal unsafe interface ISwiftProtocol
+    {
+       public abstract void* WitnessTable(ISwiftType type);
+    }
+    // <summary>
+    // Represents Swift UnsafePointer in C#.
+    // </summary>
+    internal readonly unsafe struct UnsafePointer<T> where T : unmanaged
+    {
+        private readonly T* _pointee;
+        public UnsafePointer(T* pointee)
+        {
+            this._pointee = pointee;
+        }
+
+        public T* Pointee => _pointee;
+
+        public static implicit operator T*(UnsafePointer<T> pointer) => pointer.Pointee;
+
+        public static implicit operator UnsafePointer<T>(T* pointee) => new(pointee);
+    }
+
+    // <summary>
+    // Represents Swift UnsafeMutablePointer in C#.
+    // </summary>
+    internal readonly unsafe struct UnsafeMutablePointer<T> where T : unmanaged
+    {
+        private readonly T* _pointee;
```

Comment:
The pointers all need some revision. Swift has very specific semantics that surround pointers that are used to ensure that they play nicely with the swift memory model. It's is not OK to initialize the pointer value to a `T *`.
Swift unmanaged pointers have the following states:
1. Unallocated
2. Allocated
3. Initialized
4. Deinitialized
5. Deallocated

On top of that, allocation and deallocation should always be owned by Swift and not by us.

Imagine this scenario in Swift:
```
public func consumePointerContents<T>(x: UnsafeMutablePointer<T>) {
    print ("consuming \(x.pointee)");
    x.deinitialize ();
    x.deallocate ();
}
```
If we call this from C# with a non-swift allocated and initialized UnsafeMutablePointer, we're in a load of trouble.

Similarly, in the mutable versions of these, `Pointee` should be get/set, but we should call swift to make the changes because the Swift runtime will take care of using the appropriate witness table methods to ensure that reference counting works properly.